### PR TITLE
add stats for stakes_cache_check_and_store_us

### PR DESCRIPTION
**Pull request by jeffwashington**
#### Problem

Work on tracking down why rewards calc and distribution takes so long.

#### Summary of Changes

Fixes #

### CODE REVIEW

**Comment by jeffwashington** [comment]
```
hash_accounts=0i store_accounts=431319i update_index=629384i handle_reclaims=913i append_accounts=0i stakes_cache_check_and_store_us=33970119i find_storage=0i num_accounts=147335i total_data=29771640i read_only_accounts_cache_entries=497092i read_only_accounts_cache_data_size=192529371i read_only_accounts_cache_hits=0i read_only_accounts_cache_misses=0i calc_stored_meta_us=18165i
hash_accounts=0i store_accounts=95158i update_index=210778i handle_reclaims=173i append_accounts=0i stakes_cache_check_and_store_us=6688695i find_storage=0i num_accounts=30395i total_data=12068464i read_only_accounts_cache_entries=497094i read_only_accounts_cache_data_size=192530075i read_only_accounts_cache_hits=497034i read_only_accounts_cache_misses=2i calc_stored_meta_us=3994i
hash_accounts=0i store_accounts=363668i update_index=428423i handle_reclaims=746i append_accounts=0i stakes_cache_check_and_store_us=34415587i find_storage=0i num_accounts=110361i total_data=22502704i read_only_accounts_cache_entries=497094i read_only_accounts_cache_data_size=192530075i read_only_accounts_cache_hits=0i read_only_accounts_cache_misses=0i calc_stored_meta_us=17167i
hash_accounts=0i store_accounts=432033i update_index=511323i handle_reclaims=843i append_accounts=0i stakes_cache_check_and_store_us=34128647i find_storage=0i num_accounts=135875i total_data=27525136i read_only_accounts_cache_entries=497094i read_only_accounts_cache_data_size=192530075i read_only_accounts_cache_hits=0i read_only_accounts_cache_misses=0i calc_stored_meta_us=17759i
hash_accounts=0i store_accounts=425558i update_index=546663i handle_reclaims=771i append_accounts=0i stakes_cache_check_and_store_us=34079986i find_storage=0i num_accounts=146234i total_data=29505744i read_only_accounts_cache_entries=497094i read_only_accounts_cache_data_size=192530075i read_only_accounts_cache_hits=0i read_only_accounts_cache_misses=0i calc_stored_meta_us=17814i
hash_accounts=0i store_accounts=1055606i update_index=607334i handle_reclaims=107895i append_accounts=511035i stakes_cache_check_and_store_us=10482389i find_storage=0i num_accounts=1009397i total_data=15596911i read_only_accounts_cache_entries=496908i read_only_accounts_cache_data_size=192471997i read_only_accounts_cache_hits=5i read_only_accounts_cache_misses=1i calc_stored_meta_us=301685i
```

**Comment by behzadnouri** [comment]
The metric seems to be reported every 1s. But the value of metric itself is 34s or more. So I guess this is summing up the time across all threads. How can we estimate what portion of `store_account` was spent updating stakes cache?

Also, I think it is better to use separate metrics for `store_account` vs `update_stakes_cache`. Right now both are adding to `stakes_cache_check_and_store_us` which doesn't say which portion comes from where.

**Comment by jeffwashington** [comment]
> The metric seems to be reported every 1s. But the value of metric itself is 34s or more. So I guess this is summing up the time across all threads. How can we estimate what portion of `store_account` was spent updating stakes cache?

Here are the existing metrics, reported once per second, system-wide, accumulated in each thread:
accounts_db_store_timings:
{
// writing to write cache in our case
store_accounts=431,319
// updating the accounts index because of the write
update_index=629,384
...
// this is the new one.
// this number is related to all the other numbers here
stakes_cache_check_and_store_us=33,970,119
...
// other stats:
num_accounts=147335
total_data=29771640
read_only_accounts_cache_entries=497092
read_only_accounts_cache_data_size=192529371
read_only_accounts_cache_hits=0
read_only_accounts_cache_misses=0
}

`stakes_cache_check_and_store_us` is directly relatable to `store_accounts` and `update_index`, the other two 'large' portions involved in:
`Bank::store_account`

`store_vote_accounts_us` is the largest part of `update_rewards_with_thread_pool_us`
`pay_validator_rewards_with_thread_pool` calls `Bank::store_account`
`Bank::store_account`:
1. calls `store_slow_cached` and on down the chain to build the rest of the `accounts_db_store_timings` metrics
2. calls `stakes_cache.check_and_store`, which updates `stakes_cache_check_and_store_us`

With the addition of this metric, I can see that the cost while we are calculating rewards will not be helped much by improvements to storing to the cache or updating the index.
The time is dominated by `stakes_cache_check_and_store_us`

We can also now see at other times during the epoch whether `stakes_cache.check_and_store` is costing anything significant.

**Comment by jeffwashington** [comment]
> Also, I think it is better to use separate metrics for `store_account` vs `update_stakes_cache`. Right now both are adding to `stakes_cache_check_and_store_us` which doesn't say which portion comes from where.

I don't see where `store_account` and `update_stakes_cache` are **both** adding to `stakes_cache_check_and_store_us`. The point of this change is to allow us to understand how much time `stakes_cache.check_and_store` is taking apart from accountsdb::store.

Perhaps you mean both are adding to: 
`bank-new_from_parent-new_epoch_timings.store_stake_accounts_us`
That is true. This metric is wall-clock time to do what I called 'store_stake_accounts'. Inside that store is `stakes_cache.check_and_store`. Timing these things separately would require refactoring `pay_validator_rewards_with_thread_pool` considerably due to its highly parallel nature.

I saw the addition of this metric as giving us visibility and relatable numbers for an expensive and time critical operation without introducing any risk. Can you help me understand your suggestion, please?

**Comment by behzadnouri** [comment]
I see there are already metrics for `accounts.store_cached` and `update_stakes_cache` here:
https://github.com/solana-labs/solana/blob/eae9a6653/runtime/src/bank.rs#L4738-L4758

The metrics are submitted below under `replay-slot-stats`:
https://github.com/solana-labs/solana/blob/eae9a6653/core/src/progress_map.rs#L91-L106

So this is adding some redundancy and partially duplicate metrics.
Aside from this redundancy, with current metrics `update_stakes_cache_us` / `store_us` gives the proportion of time of store op which was spent to update stakes cache which is very helpful to make sense of the raw numbers. But with this new metric, I am still not sure what is the equivalent ratio which we should look at.

In other words, how to make sense of this number:
```
stakes_cache_check_and_store_us=33,970,119
```
i.e what is total time that 33 seconds of that was spent to update stakes cache? i.e. the "ratio" I am referring to above.

**Comment by jeffwashington** [comment]
> i.e what is total time that 33 seconds of that was spent to update stakes cache? i.e. the "ratio" I am referring to above.

I believe these are all related:
// writing to write cache in our case
store_accounts=431,319
// updating the accounts index because of the write
update_index=629,384
...
// this is the new one.
// this number is related to all the other numbers here
stakes_cache_check_and_store_us=33,970,119

So, these numbers are all related. During the intense period of rewards calc, the new metric is reporting 33s is spent on stakes_cache_check_and_store. ~1s combined is spent on "store_accounts", which looks to measure time inside `AccountsDb::store_accounts_to`. Perhaps the metric name "store_accounts" is confusing? Maybe it should be renamed "store_accounts_to". `AccountsDb::store_accounts_to` runs prior to `stakes_cache_check_and_store` in the `Bank::store_account` rewards case we're looking at.
The machine I gathered metrics on is a bare metal 48 core machine. Could we have 48 cores available for 1s and spend 33s in `stakes_cache_check_and_store`? And 1s combined writing to the write cache (`AccountsDb::store_accounts_to`) and updating the index (`'update_index`)? Seems plausible.

**Comment by jeffwashington** [comment]
> I see there are already metrics for `accounts.store_cached` and `update_stakes_cache` here: https://github.com/solana-labs/solana/blob/eae9a6653/runtime/src/bank.rs#L4738-L4758
> 
> The metrics are submitted below under `replay-slot-stats`: https://github.com/solana-labs/solana/blob/eae9a6653/core/src/progress_map.rs#L91-L106

This metric is captured during `Bank::commit_transactions`.  Yes, that does also call `Bank::update_stakes_cache`. I added a github code comment to remove the redundancy.
the codepath of `commit_transactions` is not run during rewards calc.
